### PR TITLE
feat: Handle Empty String from HTML Forms

### PR DESCRIPTION
**Pull request by EnriqueL8**
This PR addresses the behaviour of the QueryDecoder with respect to parameters that have an empty string value (ie. `&key1=&key2=`).  The behaviour for types that are decoded from an empty string value is as follows:
- Any Optional type (including `String?`) defaults to `nil`
- Non-optional String successfully decodes to `""`
- Non-optional Bool decodes to `false`
- All other non-optional types throw a decoding error

### CODE REVIEW

File: Sources/KituraContracts/CodableQuery/Extensions.swift [hunk 1]
```diff
@@ -82,7 +82,7 @@ extension String {
 
     /// Converts the given String to a Bool?.
     public var boolean: Bool? {
-        return Bool(self)
+        return self != "" ? Bool(self) : false
```
Comment (ianpartridge) [inline Sources/KituraContracts/CodableQuery/Extensions.swift]: Can we use `isEmpty()` please?

File: Tests/KituraContractsTests/QueryCoderTests.swift [hunk 2]
```diff
@@ -142,20 +144,21 @@ class QueryCoderTests: XCTestCase {
...
+    let expectedQueryString = "?boolField=true&intArray=1%2C2%2C3&stringField=&optionalStringField=&intField=23&dateField=2017-12-07T21:42:06%2B0000&nested=%7B\"nestedStringField\":\"nested%20string\"%2C\"nestedIntField\":333%7D"
 
     let expectedDateStr = "2017-10-31T16:15:56+0000"
     let expectedDate = Coder().dateFormatter.date(from: "2017-10-31T16:15:56+0000")!
 
     let expectedMyQuery = MyQuery(boolField: true,
                                   intField: 23,
                                   optionalIntField: nil,
-                                  stringField: "a string",
+                                  stringField: "",
```
Comment (ianpartridge) [inline Tests/KituraContractsTests/QueryCoderTests.swift]: Instead of changing stringField to be the empty string, why not add anotherStringField so we test both the empty and non-empty cases.

**Review by djones6 — APPROVED** [review]
LGTM!

**Review by ianpartridge — APPROVED** [review]

File: Sources/KituraContracts/CodableQuery/QueryDecoder.swift [hunk 3]
```diff
@@ -29,6 +29,14 @@ import LoggerAPI
      return
  }
  ````
+ 
+ ### Decoding The Empty String: ###
+ HTML forms send the empty string ("") in query parameters when a field is empty (i.e. &key1=&key2=).
```
Comment (djones6) [inline Sources/KituraContracts/CodableQuery/QueryDecoder.swift]: I don't like the opening sentence.  We could interpret the absence of a value as an empty string, but I would prefer something like:

> ### Decoding empty values
> When an HTML form is sent with an empty or unchecked field, the corresponding key/value pair is sent with an empty value (i.e. `&key1=&key2=`).
> The mapping of empty values to Swift types is performed by `QueryDecoder` as follows:
(etc)

File: Sources/KituraContracts/Contracts.swift [hunk 4]
```diff
@@ -500,7 +500,28 @@ public extension RequestError {
...
+    let id: Int
+ }
+ router.get("/user") { (query: Query, respondWith: (User?, RequestError?) -> Void) in
+     guard let user: User = userArray[query.id] else {
+        return respondWith(nil, .notFound)
+     }
+     respondWith(user, nil)
+ }
+ ```
+ ### Decoding The Empty String: ###
```
Comment (djones6) [inline Sources/KituraContracts/Contracts.swift]: change to match amendments as above

File: Sources/KituraContracts/Contracts.swift [hunk 5]
```diff
@@ -500,7 +500,28 @@ public extension RequestError {
 }
 
 /**
- An identifier for a query parameter object.
+ An object that conforms to QueryParams is identified as being decodable from URLEncoded data.
+ This is used by Codable routes to identify that an object should be initalized by decoding the query parameters
+ using the `QueryDecoder`.
+ ### Usage Example: ###
```
Comment (djones6) [inline Sources/KituraContracts/Contracts.swift]: - remove trailing ###
- add back-ticks around the code elements
- identify -> indicate
- initalized -> initialized
...actually I'd prefer this sentence:
> This can be applied to a Codable route to define the names and types of the expected query parameters, and provide type-safe access to their values. The `QueryDecoder` is used to decode the URL encoded parameters into an instance of the conforming type.

File: Sources/KituraContracts/CodableQuery/QueryDecoder.swift (hunk 3 above)
Comment (djones6) [reply Sources/KituraContracts/CodableQuery/QueryDecoder.swift]: (also, back-ticks around code elements)
